### PR TITLE
Use fsspec as backend for data stores

### LINKED ISSUE #446: Use fsspec as backend for data store implementations
**Is your feature request related to a problem? Please describe.**

The data store implementations for the S3, directory, and memory have a lot of code in common but at the same time include some differences that involve dedicated attention, maintenance and testing effort (see also #405). Actually, the three implementations could be reduced to a single implementation by refactoring out the S3, directory, and memory as different "file systems".

Other implementation may come, and I'm thinking of HDFS that we use for our Hadoop-based EO data processing system or the Redis key-value store. 
 
**Describe the solution you'd like**

Reimplement S3, directory, and memory so that they use the [fsspec](https://filesystem-spec.readthedocs.io/) implementations "s3", "file", "memory". 

fsspec perfectly fits into the xcube data store framework:

- fsspec/s3fs is already used in S3 data store implementation therefore it is already an xcube dependency;
- fsspec implementations are configured in a consistent manner very much in line with the xcube data stoire framework;
- fsspec provides buffering, caching, supports asynchronous I/O, later also supports transactions.
- fsspec already has many implementations available (also for HDFS and Redis) which then could all be reused by xcube. Many of them implement asynchronous I/O;

**Describe alternatives you've considered**

None.

**Additional context**

### PR DESCRIPTION
**Pull request by forman**
Brand new implementation of fsspec-based data stores, openers, and writers.
Closes #446 and supports #499, #500.

The implementations of the default data stores `s3`, `directory`, and `memory` have been replaced entirely by a new implementation that utilize the [fsspec](https://filesystem-spec.readthedocs.io/) Python package. 
The preliminary filesystem-based data stores are now `s3`, `file`, and `memory`. All share a common implementations and tests. Others filesystem-based data stores can be added easily and will follow soon, for example `hdfs`.

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] ~New/modified features documented in `docs/source/*`~
* [x] Changes documented in `CHANGES.md`
* [x] AppVeyor CI passes
* [x] Test coverage remains or increases (target 100%)

Remember to close associated issues after merge!

### CODE REVIEW

**Comment by forman** [comment]
FYI @TonioF please review when you're back
FYI @dzelge @pont-us I'll present this work tomorrow

**Review by dzelge — APPROVED** [review]
I approve, but cannot review 48 files comprehenslively

File: xcube/core/store/typespecifier.py [hunk 1]
```diff
@@ -125,14 +125,17 @@ def is_satisfied_by(self, other: Union[str, "TypeSpecifier"]) -> bool:
...
+        if type_specifier is None:
+            return TYPE_SPECIFIER_ANY
         if isinstance(type_specifier, TypeSpecifier):
             return type_specifier
         if isinstance(type_specifier, str):
             return cls.parse(type_specifier)
-        raise TypeError('type_specifier must be of type '
-                        '"str" or "TypeSpecifier"')
+        raise TypeError('type_specifier must be None or a '
+                        ' str or TypeSpecifier')
```
Comment (TonioF) [inline xcube/core/store/typespecifier.py:138]: Remove blank

File: xcube/core/store/fs/impl/dataset.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                              {'time': 1, 'y': 512, 'x': 512}],
+                    additional_properties=True,
+                ),
+                decode_cf=JsonBooleanSchema(
+                    description='Whether to decode these variables,'
+                                ' assuming they were saved according to'
+                                ' CF conventions.',
+                    default=True,
+                ),
+                mask_and_scale=JsonBooleanSchema(
```
Comment (TonioF) [inline xcube/core/store/fs/impl/dataset.py:77]: Shouldn't these be two parameters?

File: xcube/core/store/fs/impl/dataset.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                        }
+                    }],
+                    additional_properties=True,
+                ),
+                consolidated=JsonBooleanSchema(
+                    description='If True, apply Zarr’s consolidate_metadata()'
+                                ' function to the store after writing.'
+                ),
+                append_dim=JsonStringSchema(
+                    description='If set, the dimension on which the'
```
Comment (TonioF) [inline xcube/core/store/fs/impl/dataset.py:152]: -> If set, data will be appended along this dimension.

File: CHANGES.md [hunk 4]
```diff
@@ -2,6 +2,14 @@
 
 ### New features
 
+* The implementations of the default data stores `s3`, `directory`, 
+  and `memory` have been replaced entirely by a new implementation
+  that utilize the [fsspec](https://filesystem-spec.readthedocs.io/) 
+  Python package. The preliminary filesystem-based data stores 
+  are now `s3`, `file`, and `memory`. All share a common implementations 
+  and tests. Others filesystem-based data stores can be added easily
+  and will follow soon, for example `hdfs`. (#446)
```
Comment (TonioF) [inline CHANGES.md:11]: Please have a look at the use of plural and singular.

**Review by TonioF — COMMENTED** [review]
The license at the start of most files is a tad too long in line 4. It needs to be replaced at some point, but this is dull.

I mostly agree with everything, but please consider my comments.

**Pull request merged**